### PR TITLE
Merge with WebAssembly/spec branch `wasm-3.0`

### PR DESCRIPTION
**Pull request by dhil**
This patch brings in the specification and reference interpreter changes from the `wasm-3.0` branch (https://github.com/WebAssembly/spec/tree/wasm-3.0).

### CODE REVIEW

**Review by rossberg — APPROVED** [review]
Rubberstamp.

Exn has been merged into wasm-3.0 a while ago, so you shouldn't need to merge that separately. (Though as long as you didn't get any merge conflicts from the latter, it shouldn't harm either. Did the 2nd merge say anything but "is already included"?)

**Comment by dhil** [comment]
> Rubberstamp.
> 
> Exn has been merged into wasm-3.0 a while ago, so you shouldn't need to merge that separately. (Though as long as you didn't get any merge conflicts from the latter, it shouldn't harm either. Did the 2nd merge say anything but "is already included"?)

I made with the 'ort' strategy. It didn't say anything about already included.
```
180 files changed, 9851 insertions(+), 1143 deletions(-)
```
If you like, I can redo the merge straight from `wasm-3.0`.

**Comment by rossberg** [comment]
Hm, to be sure nothing weird has happened, can you diff this branch against upstream/wasm-3.0? AFAICS, the diff should be empty if everything is correct.

**Comment by dhil** [comment]
This is the diff https://gist.github.com/dhil/01e35c5ffaaf1e9513350010a73520d2

TL;DR: explainer documents and code examples.

**Comment by dhil** [comment]
I can just roll back to HEAD and apply the changes from `wasm-3.0`. It is no problem.

**Comment by rossberg** [comment]
Perfect, thanks!

**Comment by slindley** [comment]
It looks to me like this PR is entirely uncontroversial and ready to go (it's just a case of bringing the stack-switching repo up-to-date). @fgmccabe please can you confirm that you're happy for it to be merged.
